File: DemoMMOUnity/Assets/Game/Net/Script/Net/CsProtocol/Protocol/Equipment/EquipmentPrimaryConfigBaseData.cs
using System.Collections.Generic;
using Newtonsoft.Json;
using ZJYFrameWork.Spring.Utils;

namespace ZJYFrameWork.Net.CsProtocol.Buffer.Protocol.Equipment
{
    public partial class EquipmentPrimaryConfigBaseData
    {
    }

    public class EquipmentPrimaryConfigBaseDataRegistration : IProtocolRegistration
    {
        public short ProtocolId()
        {
            return 208;
        }

        public void Write(ByteBuffer buffer, IPacket packet)
        {
            if (packet == null)
            {
                return;
            }

            var packetData = (EquipmentPrimaryConfigBaseData)packet;
            var json = JsonConvert.SerializeObject(packetData);
            buffer.WriteString(json);
        }

        public IPacket Read(ByteBuffer buffer)
        {
            var json = StringUtils.BytesToString(buffer.ToBytes());
            var packet = EquipmentPrimaryConfigBaseData.valueOf();
            if (string.IsNullOrEmpty(json))
            {
                return packet;
            }

            var jsonDict = JsonConvert.DeserializeObject<Dictionary<object, object>>(json);
            foreach (var (key, value) in jsonDict)
            {
                var keyString = key.ToString();
                switch (keyString)
                {
                    case "id":
                    {
                        packet.id = int.Parse(value.ToString());
                    }
                        break;
                    case "primaryQuality":
                    {
                        packet.primaryQuality = int.Parse(value.ToString());
                    }
                        break;
                    case "growthPosInt":
                    {
                        packet.growthPosInt = int.Parse(value.ToString());
                    }
                        break;
                    case "growthPosName":
                    {
                        packet.growthPosName = value.ToString();
                    }
                        break;
                    case "primaryGrowthInts":
                    {
                        packet.primaryGrowthInts = value.ToString();
                    }
                        break;
                    case "primaryGrowthMaxInt":
                    {
                        packet.primaryGrowthMaxInt = value.ToString();
                    }
                        break;
                    case "primaryGrowthName":
                    {
                        packet.primaryGrowthName = value.ToString();
                    }
                        break;
                }
            }

            return packet;
        }
    }
}

File: DemoMMOUnity/Lua/application/app/net/webscoket/luaProtocol/euipment/equipmentPrimaryConfigBaseData.lua
---@class EquipmentPrimaryConfigBaseData
local EquipmentPrimaryConfigBaseData = {}


---@param id integer
---@param primaryQuality integer
---@param growthPosInt integer
---@param primaryGrowthName string
---@param primaryGrowthInts string
---@param primaryGrowthMaxInt string
---@param growthPosName string
function EquipmentPrimaryConfigBaseData:new(
    id,
    primaryQuality,
    growthPosInt,
    primaryGrowthName,
    primaryGrowthInts,
    primaryGrowthMaxInt,
    growthPosName)
    local obj = {
        id = id,
        primaryQuality = primaryQuality,
        growthPosInt = growthPosInt,
        primaryGrowthName = primaryGrowthName,
        primaryGrowthInts = primaryGrowthInts,
        primaryGrowthMaxInt = primaryGrowthMaxInt,
        growthPosName = growthPosName
    }
    setmetatable(obj, self)
    self.__index = self
    return obj
end

function EquipmentPrimaryConfigBaseData:protocolId()
    return 208
end

function EquipmentPrimaryConfigBaseData:write(buffer, packet)
    if packet == nil then
        return
    end
    local data = packet or EquipmentPrimaryConfigBaseData
    local message = {
        protocolId = data.protocolId(),
        packet = data
    }
    local jsonStr = JSON.encode(message)
    buffer:writeString(jsonStr)
end

function EquipmentPrimaryConfigBaseData:read(buffer)
    local jsonString = buffer:readString()
    ---字节读取器中存放字符
    ---@type {protocolId:number,packet:{id:integer,primaryQuality:integer,growthPosInt:integer,primaryGrowthName:string,primaryGrowthInts:string,primaryGrowthMaxInt:string,growthPosName:string}}
    local data = JSON.decode(jsonString)
    return EquipmentPrimaryConfigBaseData:new(
        data.packet.id, data.packet.primaryQuality, data.packet.growthPosInt, data.packet.primaryGrowthName, data.packet.primaryGrowthInts,
        data.packet.primaryGrowthMaxInt, data.packet.growthPosName
    )
end

function EquipmentPrimaryConfigBaseData:readData(buffer)
    local jsonString = buffer:readString()
    ---字节读取器中存放字符
    ---@type {id:integer,primaryQuality:integer,growthPosInt:integer,primaryGrowthName:string,primaryGrowthInts:string,primaryGrowthMaxInt:string,growthPosName:string}
    local data = JSON.decode(jsonString)
    return EquipmentPrimaryConfigBaseData:new(
        data.id, data.primaryQuality, data.growthPosInt, data.primaryGrowthName, data.primaryGrowthInts,
        data.primaryGrowthMaxInt, data.growthPosName
    )
end

return EquipmentPrimaryConfigBaseData
